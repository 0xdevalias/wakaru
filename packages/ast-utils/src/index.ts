export { isFunctionExpression } from './isFunctionExpression'
export { isTopLevel } from './isTopLevel'
export { renameFunctionParameters } from './renameFunctionParameters'
export { getTopLevelStatements } from './program'
export { pruneComments } from './pruneComments'
export { wrapAstTransformation } from './wrapAstTransformation'

export * from './comments'
export * from './identifier'
export * from './imports'
export * from './isIIFE'
export * from './exports'
export * from './scope'
export * from './types'
export * from './object'
export * from './position'
export * from './reference'
