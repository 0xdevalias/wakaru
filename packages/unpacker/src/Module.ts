import type { Collection } from 'jscodeshift'

export class Module {
    id: string | number
    ast: Collection
    isEntry: boolean

    constructor(id: string | number, ast: Collection, isEntry = false) {
        this.id = id
        this.ast = ast
        this.isEntry = isEntry
    }
}
