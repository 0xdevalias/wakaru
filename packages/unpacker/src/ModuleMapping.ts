export type ModuleMapping = Record<number | string, string>
