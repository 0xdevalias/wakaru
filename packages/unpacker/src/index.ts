export type { Module } from './Module'
export type { ModuleMapping } from './ModuleMapping'

export * from './unpack'
