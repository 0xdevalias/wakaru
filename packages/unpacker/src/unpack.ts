import jscodeshift from 'jscodeshift'

// @ts-expect-error - no types
import getParser from 'jscodeshift/src/getParser'
import { getModulesFromBrowserify } from './extractors/browserify'
import { getModulesFromWebpack } from './extractors/webpack'
import type { Module } from './Module'
import type { ModuleMapping } from './ModuleMapping'

/**
 * Unpacks the given source code from supported bundlers.
 */

export function unpack(sourceCode: string): {
    modules: Module[]
    moduleIdMapping: ModuleMapping
} | null {
    const parser = getParser()
    const j = jscodeshift.withParser(parser)
    const root = j(sourceCode)

    const result = getModulesFromWebpack(j, root)
        || getModulesFromBrowserify(j, root)

    if (!result) {
        console.error('Failed to locate modules')
        return null
    }

    const { modules, moduleIdMapping } = result

    return {
        modules: [...modules],
        moduleIdMapping,
    }
}
