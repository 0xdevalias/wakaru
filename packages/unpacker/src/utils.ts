import type { ASTPath, ArrowFunctionExpression, Collection, ExpressionStatement, FunctionExpression, JSCodeshift, Node, Statement } from 'jscodeshift'
import prettier from 'prettier/standalone'
import babelParser from 'prettier/parser-babel'

export function isTopLevel(j: JSCodeshift, node: ASTPath<Node>): boolean {
    return j.Program.check(node.parentPath.node)
}

export function pruneComments(j: JSCodeshift, collection: Collection<any>): void {
    // @ts-expect-error - Comment type is wrong
    collection.find(j.Comment).forEach(path => path.prune())
}

export function renameFunctionParameters(j: JSCodeshift, node: FunctionExpression | ArrowFunctionExpression, parameters: string[]): void {
    node.params.forEach((param, index) => {
        if (param.type === 'Identifier') {
            j(node)
                .find(j.Identifier, { name: param.name })
                .filter(path => path.scope.node === node)
                .forEach((path) => {
                    path.node.name = parameters[index]
                })
        }
    })
}

export function isIIFE(node: Statement): node is ExpressionStatement {
    if (node.type !== 'ExpressionStatement') return false
    const expression = (node as ExpressionStatement).expression
    if (expression.type !== 'CallExpression') return false
    const callee = expression.callee
    return callee.type === 'FunctionExpression'
        || callee.type === 'ArrowFunctionExpression'
}

export function prettierFormat(code: string) {
    return prettier.format(code, {
        parser: 'babel',
        plugins: [babelParser],
    })
}
