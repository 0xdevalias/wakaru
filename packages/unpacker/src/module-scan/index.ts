import { ExportManager } from '@wakaru/ast-utils/exports'
import { ImportManager } from '@wakaru/ast-utils/imports'
import { postScanBabelRuntime, scanBabelRuntime } from './babel-runtime'
import type { Module } from '../Module'
import type { Collection, JSCodeshift } from 'jscodeshift'

export function scanImports(j: JSCodeshift, module: Module & { root: Collection }) {
    const importManager = new ImportManager()
    importManager.collectEsModuleImport(j, module.root)
    importManager.collectCommonJsImport(j, module.root)
    module.import = importManager.getModuleImports()
}

export function scanExports(j: JSCodeshift, module: Module & { root: Collection }) {
    const exportManager = new ExportManager()
    exportManager.collectEsModuleExport(j, module.root)
    exportManager.collectCommonJsExport(j, module.root)
    module.export = exportManager.toJSON()
}

export function scanRuntime(j: JSCodeshift, module: Module & { root: Collection }) {
    scanBabelRuntime(j, module)
}

export function postScanRuntime(j: JSCodeshift, modules: (Module & { root: Collection })[]) {
    postScanBabelRuntime(j, modules)
}
