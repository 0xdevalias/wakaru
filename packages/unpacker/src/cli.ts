#!/usr/bin/env node

/* eslint-disable no-console */
import * as path from 'node:path'
import process, { hrtime } from 'node:process'
import fsa from 'fs-extra'
import * as globby from 'globby'
import c from 'picocolors'
import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'
import { version } from '../package.json'
import { unpack } from './unpack'

type LogLevel = 'error' | 'warn' | 'info' | 'debug' | 'silent'

// eslint-disable-next-line no-unused-expressions
yargs(hideBin(process.argv))
    .scriptName('@wakaru/unpacker')

    .help()
    .showHelpOnFail(true)
    .alias('h', 'help')

    .version('version', version)
    .alias('v', 'version')

    .option('log-level', {
        type: 'string',
        default: 'info',
        choices: ['error', 'warn', 'info', 'debug', 'silent'],
        describe: 'change the level of logging for the CLI.',
    })

    .usage('Usage: $0 <files...> [options]')
    .command(
        '* <files...>',
        'Unminify your bundled code',
        args => args
            .option('output', {
                alias: 'o',
                describe: 'specify the output directory (default: out/)',
                type: 'string',
                default: 'out/',
            })
            .option('force', {
                alias: 'f',
                describe: 'force overwrite output directory',
                type: 'boolean',
                default: false,
            })
            .positional('files', {
                describe: 'File paths to process (supports glob patterns)',
                type: 'string',
                array: true,
            })
            .help(),
        async (args) => {
            await unpacker(
                args.files ?? [],
                args.output,
                args.force,
                args.logLevel as LogLevel,
            )
        },
    )
    .argv

async function unpacker(
    paths: string[],
    output: string,
    force: boolean,
    logLevel: LogLevel,
) {
    const cwd = process.cwd()
    const globbyPaths = paths
        .map(p => path.normalize(p))
        .map(p => p.replace(/\\/g, '/'))
    const resolvedPaths = await globby.default(globbyPaths, {
        cwd,
        absolute: true,
        ignore: [path.join(cwd, '**/node_modules/**')],
    })

    // Check if any paths are outside of the current working directory
    for (const p of resolvedPaths) {
        if (!isPathInside(cwd, p)) {
            throw new Error(`File path ${c.green(path.relative(cwd, p))} is outside of the current working directory. This is not allowed.`)
        }
    }

    const outputDir = path.resolve(cwd, output)

    if (await fsa.exists(outputDir)) {
        if (!force) {
            throw new Error(`Output directory already exists at ${c.green(path.relative(cwd, outputDir))}. Pass ${c.yellow('--force')} to overwrite.`)
        }
    }
    await fsa.ensureDir(outputDir)

    for (const p of resolvedPaths) {
        const start = hrtime()

        const source = await fsa.readFile(p, 'utf-8')
        const { modules, moduleIdMapping } = unpack(source)

        const folder = resolvedPaths.length > 1 ? path.basename(p, path.extname(p)) : ''
        const outputFolder = path.join(outputDir, folder)
        for (const mod of modules) {
            const filename = moduleIdMapping[mod.id] ?? `module-${mod.id}.js`
            const outputPath = path.join(outputFolder, filename)
            await fsa.ensureDir(path.dirname(outputPath))
            await fsa.writeFile(outputPath, mod.code, 'utf-8')
        }

        if (logLevel !== 'silent') {
            const end = hrtime(start)
            const elapsed = end[0] * 1e9 + end[1]
            const formattedElapsed = (elapsed / 1e6).toLocaleString('en-US', { maximumFractionDigits: 1 })
            console.log(`Generated ${c.green(modules.length)} modules from ${c.green(path.relative(cwd, p))} to ${c.green(path.relative(cwd, outputFolder))} ${c.dim(`(${formattedElapsed}ms)`)}`)
        }
    }
}

/**
 * Check if base path contains target path
 */
function isPathInside(base: string, target: string): boolean {
    const relative = path.relative(base, target)
    return !relative.startsWith('..') && !path.isAbsolute(relative)
}
