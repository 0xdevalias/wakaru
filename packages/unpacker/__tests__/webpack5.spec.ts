import { readFile } from 'node:fs/promises'
import { describe, expect, it } from 'vitest'
import { unpack } from '../src'

describe('Webpack 5', () => {
    it('testcases/webpack5', async () => {
        const source = await readFile('../../testcases/webpack5/dist/index.js', 'utf8')
        const result = unpack(source)
        if (!result) throw new Error('Failed to unpack')

        expect(result.moduleIdMapping).toMatchSnapshot()

        expect(result.modules.size).toBe(5)

        const modules = [...result.modules.values()]
            .map(({ id, isEntry, code }) => ({ id, isEntry, code }))
        expect(modules).toMatchSnapshot()
    })
})
