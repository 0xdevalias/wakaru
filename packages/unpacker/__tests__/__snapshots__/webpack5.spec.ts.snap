// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Webpack 5 > testcases/webpack5 1`] = `{}`;

exports[`Webpack 5 > testcases/webpack5 2`] = `
[
  {
    "code": "/* harmony export */ require.d(exports, {
  /* harmony export */ default: () => /* binding */ m1,
  /* harmony export */
});
function m1() {
  console.log("m1");
}

/***/
",
    "id": "./src/1.js",
    "isEntry": false,
  },
  {
    "code": "/* harmony export */ require.d(exports, {
  /* harmony export */ A: () => /* binding */ A,
  /* harmony export */
});
class A {
  constructor() {
    this.label = "a";
  }

  print() {
    console.log("a", this.version);
  }
}

/***/
",
    "id": "./src/a.js",
    "isEntry": false,
  },
  {
    "code": "/* harmony export */ require.d(exports, {
  /* harmony export */ default: () =>
    /* export default binding */ __WEBPACK_DEFAULT_EXPORT__,
  /* harmony export */ version: () => /* binding */ version,
  /* harmony export */
});
const version = "1.0.0";

/* harmony default export */ function __WEBPACK_DEFAULT_EXPORT__() {
  console.log("b", version);
}

/***/
",
    "id": "./src/b.js",
    "isEntry": false,
  },
  {
    "code": "/* harmony export */ require.d(exports, {
  /* harmony export */ getC: () => /* binding */ getC,
  /* harmony export */
});
/* harmony import */ var _b_js__WEBPACK_IMPORTED_MODULE_0__ = require(/*! ./b.js */ "./src/b.js");

const getC = async () => {
  console.log("c.a", _b_js__WEBPACK_IMPORTED_MODULE_0__.version);
  const result = await fetch("https://jsonplaceholder.typicode.com/todos/1");
  const json = await result.json();
  return json;
};

/***/
",
    "id": "./src/c.js",
    "isEntry": false,
  },
  {
    "code": "/*!**********************!*\\
  !*** ./src/index.js ***!
  \\**********************/
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _a_js__WEBPACK_IMPORTED_MODULE_0__ =
  __webpack_require__(/*! ./a.js */ "./src/a.js");
/* harmony import */ var _b_js__WEBPACK_IMPORTED_MODULE_1__ =
  __webpack_require__(/*! ./b.js */ "./src/b.js");
/* harmony import */ var _c_js__WEBPACK_IMPORTED_MODULE_2__ =
  __webpack_require__(/*! ./c.js */ "./src/c.js");
/* harmony import */ var _1_js__WEBPACK_IMPORTED_MODULE_3__ =
  __webpack_require__(/*! ./1.js */ "./src/1.js");

console.log(
  _b_js__WEBPACK_IMPORTED_MODULE_1__.version,
  _a_js__WEBPACK_IMPORTED_MODULE_0__.A
);
(0, _b_js__WEBPACK_IMPORTED_MODULE_1__["default"])();
(0, _c_js__WEBPACK_IMPORTED_MODULE_2__.getC)().then(console.log);

// const M1 = await import('./1.js')
(0, _1_js__WEBPACK_IMPORTED_MODULE_3__["default"])();
",
    "id": "entry.js",
    "isEntry": true,
  },
]
`;
